src/adversarial/neurify.jl

```diff
@@ -0,0 +1,306 @@
+"""
+    Neurify(max_iter::Int64, tree_search::Symbol)
+
+Neurify combines symbolic reachability analysis with constraint refinement to minimize over-approximation of the reachable set.
+
+# Problem requirement
+1. Network: any depth, ReLU activation
+2. Input: AbstractPolytope
+3. Output: AbstractPolytope
+
+# Return
+`CounterExampleResult` or `ReachabilityResult`
+
+# Method
+Symbolic reachability analysis and iterative interval refinement (search).
+- `max_iter` default `10`.
+
+# Property
+Sound but not complete.
+
+# Reference
+[S. Wang, K. Pei, J. Whitehouse, J. Yang, and S. Jana,
+"Efficient Formal Safety Analysis of Neural Networks,"
+*CoRR*, vol. abs/1809.08098, 2018. arXiv: 1809.08098.](https://arxiv.org/pdf/1809.08098.pdf)
+[https://github.com/tcwangshiqi-columbia/Neurify](https://github.com/tcwangshiqi-columbia/Neurify)
+"""
+
+@with_kw struct Neurify
+    max_iter::Int64     = 100
+    tree_search::Symbol = :DFS # only :DFS/:BFS allowed? If so, we should assert this.
+end
+
+# Why can "interval" be an AbstractPolytope?
+struct SymbolicInterval{F<:AbstractPolytope}
+    Low::Matrix{Float64}
+    Up::Matrix{Float64}
+    interval::F
+end
+
+# Data to be passed during forward_layer
+struct SymbolicIntervalGradient
```

Comment:
Is there a conceptual difference between `SymbolicIntervalGradient` and `SymbolicIntervalMask` from reluval? Do `LΛ/UΛ` represent the same relu mask still? If the only different is `r`, I'm wondering whether a refactoring will help consolidate. Does the original paper do something similar to this, or is their approach different?

Reply:
Yes, `SymbolicIntervalGradient` records gradients as real numbers which can be anything between 0 and 1 (consider relexed_ReLU), `SymbolicIntervalMask` records binary 0/1 gradients as integers. We can actually merge the two definitions. And we probably don't need to record `r`. We can compute it from other variables.